File: questions/kmeans/ts/dataviz.ts
/**
 * Looking at an answer: the ASCII scatter plot and the one-line-per-cluster dump.
 *
 * `dataviz` is the name the brief gives the candidate, and the pad puts this module beside
 * their solution rather than pasting it into it.
 *
 *     show({ points })                     -> every point is '·'
 *     show(clusters)                       -> one mark per group, in list order
 *     show(clusters, centroids)            -> centroids overlaid as their group's digit
 *
 * `clusters` is a list of groups, so a single group is `show({ points })` — passing one
 * bare list of points is the easy mistake and throws rather than plotting nonsense. The
 * object form is what Python spells `show(points=pts)`; TypeScript has no keyword args.
 *
 * A group is any `Iterable<Point>` — an array, a generator, whatever. `Iterable` rather
 * than `Array` is deliberate and the opposite of kmeans(): show() makes exactly one pass
 * and materializes, so a generator is safe here in a way it is not for a multi-pass
 * algorithm.
 *
 * Non-finite coordinates are dropped and counted rather than thrown on, so a half-broken
 * solution still draws something.
 *
 * width/height of 0 mean auto: both come from the terminal, leaving room for the borders
 * and the prompt. The data is then stretched to fill that box on each axis independently,
 * so the result is a topology view rather than a scale drawing. Pass them explicitly for
 * a fixed size.
 *
 * Ported from `py/dataviz.py`. The one thing the port cannot carry over is
 * the Point/Centroid int/float split: it is documentation only, TypeScript has one number
 * type.
 */

/** Python's `round`: half-to-even, unlike JS's half-up `Math.round`. */
function round(x: number): number {
  const floor = Math.floor(x);
  const frac = x - floor;
  if (frac > 0.5) return floor + 1;
  if (frac < 0.5) return floor;
  return floor % 2 === 0 ? floor : floor + 1;
}

export const MARKS = "●▲■◆★✚✦❖"; // if your terminal misaligns these, use "oxv+*#@%"
const UNLABELED = "·";
const BLANK = " ";

/** Data points are integral (pixels, counts, ages) — by convention, not by type. */
export type Point = readonly [number, number];
/** A centroid is a mean, and rarely integral. */
export type Centroid = readonly [number, number];

/** (row, col) into the character grid, flattened to `row * width + col`. */
type Cell = number;

export interface ShowBox {
  height?: number;
  width?: number;
  title?: string;
}

const XY = 2; // a Point, a Centroid and a Pair are all 2-arrays; only their contents differ

/** A cluster as cluster() returns it: its centroid, then its points. */
export type Pair = [Centroid, Iterable<Point>];

/**
 * Every shape the first argument may take. A group is Iterable rather than Array on purpose —
 * show() makes exactly one pass, so a generator is safe here.
 */
export type Clusters = Map<Centroid, Iterable<Point>> | Point[] | Iterable<Point>[] | Pair[];

/** The single-group call: `show({ points })`, TypeScript's stand-in for a keyword arg. */
export interface ShowSpec extends ShowBox {
  points: Iterable<Point>;
  centroids?: Iterable<Centroid>;
}

/** Split points into the plottable ones and a count of the rest. */
function finite(points: Iterable<Centroid>): [Centroid[], number] {
  const usable: Centroid[] = [];
  let dropped = 0;
  for (const [x, y] of points) {
    if (Number.isFinite(x) && Number.isFinite(y)) usable.push([x, y]);
    else dropped++;
  }
  return [usable, dropped];
}

/** Grid size, defaulting to the terminal with room for borders and the prompt. */
function terminalBox(width: number, height: number): [number, number] {
  const cols = process.stdout.columns || 80;
  const rows = process.stdout.rows || 24;
  return [
    width || Math.max(20, Math.min(120, cols - 2)), // 2 columns for the │ borders
    height || Math.max(5, Math.min(60, rows - 4)), // 2 rules, a title, a prompt
  ];
}

/** Map data coordinates onto grid cells, stretching each axis to fill the box. */
function projection(points: Centroid[], width: number, height: number): (point: Centroid) => Cell {
  // one pass rather than Math.min(...xs): spreading an array of arguments throws RangeError
  // past ~65k of them, and Python's min() over a generator has no such ceiling to match.
  let x0 = Infinity;
  let x1 = -Infinity;
  let y0 = Infinity;
  let y1 = -Infinity;
  for (const [x, y] of points) {
    if (x < x0) x0 = x;
    if (x > x1) x1 = x;
    if (y < y0) y0 = y;
    if (y > y1) y1 = y;
  }
  const spanX = x1 - x0 || 1;
  const spanY = y1 - y0 || 1;

  return ([x, y]) => {
    const col = round(((x - x0) / spanX) * (width - 1));
    const row = round(((y1 - y) / spanY) * (height - 1)); // flip y: row 0 is the top
    return Math.max(0, Math.min(height - 1, row)) * width + Math.max(0, Math.min(width - 1, col));
  };
}

const isPoint = (value: unknown): value is Point =>
  Array.isArray(value) && value.length === XY && value.every((v) => typeof v === "number");

/** A [centroid, its points] pair — what cluster() returns, one per cluster. */
const isPair = (value: unknown): value is Pair =>
  Array.isArray(value) && value.length === XY && isPoint(value[0]) && !isPoint(value[1]);

/**
 * Work out which call shape this is. See the module docstring for all of them.
 *
 * Each is told apart by its first element and they cannot collide: a Point is a 2-array of
 * numbers, a pair is a 2-array whose first element is one, and a group of points is neither.
 */
function resolve(first: Clusters, centroids: Iterable<Centroid> | undefined): [Iterable<Point>[], Iterable<Centroid>] {
  if (first instanceof Map) return [[...first.values()], [...first.keys()]];

  const items = [...first];
  if (items.length === 0) return [[], centroids ?? []];
  if (isPoint(items[0])) return [[items as Point[]], centroids ?? []]; // one bare group
  if (isPair(items[0])) {
    const pairs = items as Pair[];
    return [pairs.map(([, pts]) => pts), pairs.map(([centroid]) => centroid)];
  }
  return [items as Iterable<Point>[], centroids ?? []];
}

/**
 * Which digit on the plot is which centroid, and the mark of the group it belongs to.
 *
 * A centroid past the last group gets no mark: the pairing is positional, so there is
 * nothing for it to name.
 */
function legend(centers: Centroid[], marks: string, groups: number): string {
  return centers
    .map(([cx, cy], index) => `(${g(cx, 4)}, ${g(cy, 4)}): ${index % 10}${index < groups ? marks[index % marks.length] : ""}`)
    .join("  ");
}

export function show(clusters: Map<Centroid, Iterable<Point>>, box?: ShowBox): void;
export function show(clusters: Pair[], box?: ShowBox): void;
export function show(points: Point[], box?: ShowBox): void;
export function show(clusters: Iterable<Point>[], centroids?: Iterable<Centroid>, box?: ShowBox): void;
// after the centroids form, so `show(groups, C)` still picks that one — a box is not iterable
export function show(clusters: Iterable<Point>[], box?: ShowBox): void;
export function show(spec: ShowSpec): void;

/** Print an ASCII scatter, one mark per group. See the module docstring. */
export function show(
  first: Clusters | ShowSpec = [],
  centroids?: Iterable<Centroid> | ShowBox,
  box: ShowBox = {},
): void {
  const spec = Array.isArray(first) || first instanceof Map ? null : first;
  // the 2-arg overloads pass a box where the 3-arg one passes centroids
  const asBox = centroids !== undefined && !(Symbol.iterator in Object(centroids)) ? (centroids as ShowBox) : undefined;
  const given = asBox ? undefined : (centroids as Iterable<Centroid> | undefined);
  const opts = spec ?? { ...(asBox ?? box) };

  const [resolved, found] = spec ? [[spec.points], spec.centroids ?? []] : resolve(first as Clusters, given);
  const groups = resolved.map(finite);
  const [centers, centroidsDropped] = finite(found);
  const dropped = centroidsDropped + groups.reduce((sum, [, n]) => sum + n, 0);
  const plotted = groups.flatMap(([group]) => group);

  if (plotted.length === 0 && centers.length === 0) {
    console.log(dropped ? `(nothing to plot — ${dropped} unusable)` : "(no points)");
    return;
  }

  const [width, height] = terminalBox(opts.width ?? 0, opts.height ?? 0);
  const cellOf = projection([...plotted, ...centers], width, height);

  // Groups overlap, so tally every mark landing in a cell and let the majority hold it.
  // Insertion order breaks ties, which is how Python's Counter.most_common(1) breaks them.
  const marks = groups.length === 1 ? UNLABELED : MARKS;
  const tally = new Map<Cell, Map<string, number>>();
  groups.forEach(([group], index) => {
    const mark = marks[index % marks.length]!;
    for (const point of group) {
      const cell = cellOf(point);
      const here = tally.get(cell) ?? new Map<string, number>();
      here.set(mark, (here.get(mark) ?? 0) + 1);
      tally.set(cell, here);
    }
  });

  const grid = Array.from({ length: height }, () => new Array<string>(width).fill(BLANK));
  for (const [cell, here] of tally) {
    let best = BLANK;
    let bestCount = 0;
    for (const [mark, count] of here) {
      if (count > bestCount) [best, bestCount] = [mark, count];
    }
    grid[Math.floor(cell / width)]![cell % width] = best;
  }
  centers.forEach((center, index) => {
    // drawn last, so a centroid wins its cell
    const cell = cellOf(center);
    grid[Math.floor(cell / width)]![cell % width] = String(index % 10);
  });

  const rule = "─".repeat(width);
  const notes = [opts.title ?? "", dropped ? `${dropped} point(s) unusable` : ""].filter(Boolean);
  if (centers.length) notes.push(legend(centers, marks, groups.length));
  console.log(`┌${rule}`);
  console.log(grid.map((row) => "│" + row.join("")).join("\n"));
  console.log(`└${rule}  ` + notes.join("  ·  "));
}

export type Cluster = [Centroid, Point[]];

/** Significant digits with trailing zeros dropped — Python's `%g` / `%.4g`. */
const g = (value: number, digits: number): string => String(Number(value.toPrecision(digits)));

const byValue = (a: readonly [number, number], b: readonly [number, number]): number => a[0] - b[0] || a[1] - b[1];

/**
 * One line per cluster: `centroid: points`.
 *
 * Sorted so two runs are diffable — cluster order and point order are not part of the
 * contract, and sorting inside kmeans() would be a misread of it.
 */
export function printClusters(clusters: readonly Cluster[]): void {
  for (const [centroid, points] of [...clusters].sort(([a], [b]) => byValue(a, b))) {
    const coords = [...points]
      .sort(byValue)
      .map(([x, y]) => `(${g(x, 6)},${g(y, 6)})`)
      .join(",");
    console.log(`(${g(centroid[0], 4)}, ${g(centroid[1], 4)}): ${coords}`);
  }
}

/** Self-test: the input shapes show() accepts. */
export function demo(): void {
  const quad: Point[] = Array.from({ length: 41 }, (_, i) => {
    const x = i - 20;
    return [x, Math.floor((x * x) / 8) - 40];
  });
  const left = quad.filter(([x]) => x < 0);
  const right = quad.filter(([x]) => x >= 0);

  show({ points: quad, width: 44, height: 8, title: "one group -> unlabeled" });
  const halves: Centroid[] = [
    [-10, -20],
    [10, -20],
  ];
  show([left, right], halves, { width: 44, height: 8, title: "two groups + centroids" });
  show([left.values(), right.values()], undefined, { width: 44, height: 8, title: "iterators — safe, show() is single-pass" });
  show([quad], [[0, NaN]], { width: 44, height: 8, title: "NaN centroid does not crash" });
  show([], undefined, { width: 44 });
}

if (import.meta.main) demo();

File: questions/kmeans/py/dataviz.py
"""Looking at an answer: the ASCII scatter plot and the one-line-per-cluster dump.

`dataviz` is the name the brief gives the candidate, and the pad puts this module beside
their solution rather than pasting it into it.


    show(pts)                     -> every point is '·'
    show(points=pts)              -> the same, spelled as a keyword
    show(groups)                  -> one mark per group, in list order
    show(groups, C)               -> centroids overlaid as their group's digit
    show(kmeans(pts, k))          -> [(centroid, its points), ...] — both at once
    show({centroid: pts, ...})    -> the same, as a mapping

Each shape is told apart by its first element and they cannot collide: a Point is a 2-tuple
of numbers, a (centroid, points) pair is a 2-tuple whose first element is one, and a group
of points is neither.

A group is any iterable of Point — a list, a generator, whatever. `Iterable` rather than
`Sequence` is deliberate and the opposite of kmeans(): show() makes exactly one pass and
materializes, so a generator is safe here in a way it is not for a multi-pass algorithm.

Non-finite coordinates are dropped and counted rather than raised on, so a half-broken
solution still draws something.

width/height of 0 mean auto: both come from the terminal, leaving room for the borders and
the prompt. The data is then stretched to fill that box on each axis independently, so the
result is a topology view rather than a scale drawing. Pass them explicitly for a fixed size.
"""

from __future__ import annotations  # so `| None` works on Python 3.9

from collections import Counter, defaultdict
from collections.abc import Iterable, Sequence
from itertools import cycle
from math import isfinite
from shutil import get_terminal_size
from typing import TYPE_CHECKING, cast, overload

if TYPE_CHECKING:
    from collections.abc import Callable

MARKS = "●▲■◆★✚✦❖"  # if your terminal misaligns these, use "oxv+*#@%"
UNLABELED = "·"
BLANK = " "


# Plain tuples — no constructor to import, nothing to convert. The int/float split is
# the domain: data points are integral (pixels, counts, ages), a centroid is a mean and
# rarely is. By the numeric tower a Point is accepted wherever a Centroid is expected,
# but not the reverse — so a mean can never be mistaken for a data point.
Point = tuple[int, int]
Centroid = tuple[float, float]

Cell = tuple[int, int]  # (row, col) into the character grid
Pair = tuple[Centroid, Sequence[Point]]  # what kmeans() returns, one per cluster

# Every shape the first positional argument may take. A group is Iterable rather than
# Sequence on purpose — show() makes exactly one pass, so a generator is safe here.
Clusters = dict[Centroid, Sequence[Point]] | Sequence[Point] | Sequence[Iterable[Point]] | Iterable[Pair]


def _finite(points: Iterable[Centroid]) -> tuple[list[Centroid], int]:
    """Split points into the plottable ones and a count of the rest."""
    usable: list[Centroid] = []
    dropped = 0
    for x, y in points:
        if isfinite(x) and isfinite(y):
            usable.append((x, y))
        else:
            dropped += 1
    return usable, dropped


def _terminal_box(width: int, height: int) -> tuple[int, int]:
    """Grid size, defaulting to the terminal with room for borders and the prompt."""
    cols, rows = get_terminal_size((80, 24))
    return (
        width or max(20, min(120, cols - 2)),  # 2 columns for the │ borders
        height or max(5, min(60, rows - 4)),  # 2 rules, a title, a prompt
    )


def _projection(points: list[Centroid], width: int, height: int) -> Callable[[Centroid], Cell]:
    """Map data coordinates onto grid cells, stretching each axis to fill the box."""
    x0, x1 = min(x for x, _ in points), max(x for x, _ in points)
    y0, y1 = min(y for _, y in points), max(y for _, y in points)
    span_x, span_y = (x1 - x0) or 1.0, (y1 - y0) or 1.0

    def cell(point: Centroid) -> Cell:
        x, y = point
        col = round((x - x0) / span_x * (width - 1))
        row = round((y1 - y) / span_y * (height - 1))  # flip y: row 0 is the top
        return max(0, min(height - 1, row)), max(0, min(width - 1, col))

    return cell


def _is_point(value: object) -> bool:
    match value:
        case (int() | float(), int() | float()):
            return True
        case _:
            return False


def _resolve(
    clusters: Clusters,
    points: Iterable[Point] | None,
    centroids: Iterable[Centroid] | None,
) -> tuple[list[Iterable[Point]], list[Centroid]]:
    """Work out which of the call shapes this is. See the module docstring for all of them.

    Each is told apart by its first element and they cannot collide: a Point is a 2-sequence of
    numbers, a (centroid, points) pair is a 2-sequence whose first element is one, and a group
    of points is neither.
    """
    given = list(centroids) if centroids is not None else []
    if points is not None:
        return [points], given
    if isinstance(clusters, dict):
        # The mapping form: keys are the centroids, values their points. Checked before the
        # match below because a dict iterates as its keys, which is not what it means here.
        # The cast is because a dict is also an Iterable of its keys, so `Iterable[Pair]`
        # matches it too and widens them.
        mapping = cast("dict[Centroid, Sequence[Point]]", clusters)
        return list(mapping.values()), list(mapping.keys())

    items = list(clusters)
    match items:
        case []:
            return [], given
        case [(int() | float(), int() | float()), *_]:
            return [cast("list[Point]", items)], given  # one bare group: what a candidate reaches for first
        # `not _is_point(group)` is what keeps a two-point group from reading as a pair
        case [(centroid, group), *_] if _is_point(centroid) and not _is_point(group):
            pairs = cast("list[Pair]", items)
            return [pts for _, pts in pairs], [centroid for centroid, _ in pairs]
        case _:
            return cast("list[Iterable[Point]]", items), given


@overload
def show(clusters: dict[Centroid, Sequence[Point]], /, *, title: str = "", height: int = 0, width: int = 0) -> None: ...
@overload
def show(clusters: Iterable[Pair], /, *, title: str = "", height: int = 0, width: int = 0) -> None: ...
@overload
def show(points: Sequence[Point], /, *, title: str = "", height: int = 0, width: int = 0) -> None: ...
@overload
def show(clusters: Sequence[Iterable[Point]], /, *, title: str = "", height: int = 0, width: int = 0) -> None: ...
@overload
def show(
    clusters: Sequence[Iterable[Point]], centroids: Iterable[Centroid], /, *, title: str = "", height: int = 0, width: int = 0
) -> None: ...
@overload
def show(*, points: Iterable[Point] | None = None, title: str = "", height: int = 0, width: int = 0) -> None: ...


def _legend(centers: list[Centroid], marks: str, groups: int) -> str:
    """Which digit on the plot is which centroid, and the mark of the group it belongs to.

    A centroid past the last group gets no mark: the pairing is positional, so there is
    nothing for it to name.
    """
    return "  ".join(
        f"({cx:.4g}, {cy:.4g}): {index % 10}{marks[index % len(marks)] if index < groups else ''}"
        for index, (cx, cy) in enumerate(centers)
    )


def show(  # ruff: ignore[too-many-arguments] — width/height/title are plotting knobs, keyword-only and defaulted
    clusters: Clusters = (),
    centroids: Iterable[Centroid] | None = None,
    *,
    points: Iterable[Point] | None = None,
    title: str = "",
    height: int = 0,
    width: int = 0,
) -> None:
    """Print an ASCII scatter, one mark per group. See the module docstring."""
    resolved, centers = _resolve(clusters, points, centroids)
    groups = [_finite(group) for group in resolved]
    centers, dropped = _finite(centers)
    dropped += sum(n for _, n in groups)
    plotted = [point for group, _ in groups for point in group]

    if not plotted and not centers:
        print(f"(nothing to plot — {dropped} unusable)" if dropped else "(no points)")
        return

    width, height = _terminal_box(width, height)
    cell_of = _projection(plotted + centers, width, height)

    # Groups overlap, so tally every mark landing in a cell and let the majority hold it.
    marks = UNLABELED if len(groups) == 1 else MARKS
    tally: defaultdict[Cell, Counter[str]] = defaultdict(Counter)
    for mark, (group, _) in zip(cycle(marks), groups):
        for point in group:
            tally[cell_of(point)][mark] += 1

    grid = [[BLANK] * width for _ in range(height)]
    for (row, col), here in tally.items():
        grid[row][col] = here.most_common(1)[0][0]
    for index, center in enumerate(centers):  # drawn last, so a centroid wins its cell
        row, col = cell_of(center)
        grid[row][col] = str(index % 10)

    rule = "─" * width
    notes = [note for note in (title, f"{dropped} point(s) unusable" if dropped else "") if note]
    if centers:
        notes.append(_legend(centers, marks, len(groups)))
    print(f"┌{rule}")
    print("\n".join("│" + "".join(row) for row in grid))
    print(f"└{rule}  " + "  ·  ".join(notes))


def print_clusters(clusters: list[tuple[Centroid, list[Point]]]) -> None:
    """One line per cluster: `centroid: points`.

    Sorted so two runs are diffable — cluster order and point order are not part of the
    contract, and sorting inside kmeans() would be a misread of it.
    """
    for centroid, pts in sorted(clusters):
        coords = ",".join(f"({x:g},{y:g})" for x, y in sorted(pts))
        cx, cy = centroid
        print(f"({cx:.4g}, {cy:.4g}): {coords}")


def _demo() -> None:
    """Self-test: the input shapes show() accepts. Entry point for `interview-k`."""
    quad: list[Point] = [(x, x * x // 8 - 40) for x in range(-20, 21)]
    left = [p for p in quad if p[0] < 0]
    right = [p for p in quad if p[0] >= 0]

    show(points=quad, width=44, height=8, title="one group -> unlabeled")
    show([left, right], [(-10.0, -20.0), (10.0, -20.0)], width=44, height=8, title="two groups + centroids")
    show([(p for p in left), (p for p in right)], width=44, height=8, title="generators — safe, show() is single-pass")
    show([quad], [(0.0, float("nan"))], width=44, height=8, title="nan centroid does not crash")
    show(width=44)

    try:
        import numpy as np
    except ImportError:
        print("(numpy absent — stdlib path fine)")
    else:
        rng = np.random.default_rng(1)
        arr = rng.normal(0, 20, (80, 2))
        pts: list[Point] = [(round(x), round(y)) for x, y in arr]  # ndarray rows -> Point
        mid = [[p for p in pts if p[0] < 0], [p for p in pts if p[0] >= 0]]
        show(mid, [(-20.0, 0.0), (20.0, 0.0)], width=44, title="from an ndarray")


if __name__ == "__main__":
    _demo()
